Adjustments to header spacing and alignment

querkmachine opened the pull request:
Adjust spacing and alignment between elements in the header and logo. A lot of the changes here are 🪄 magic numbers 🔢 that aim to achieve the spacing desired by the brand guidelines, rather than good old reliable design system numbers. Here be dragons. 

Based on top of #5857 so that things are measured against the most recent logo graphic. Closes alphagov/govuk-design-system#4612 and alphagov/govuk-design-system#4615.

## Changes
- Made the space between the logo and product name 7 pixels (when inline) and 2 pixels (when stacked).
- Increased the font size of the product name to 31 pixels.
  - Because of the spacing inherent in typeface glyphs, this results in type that is visually 22 pixels from baseline to cap height—the same height as the GOV.UK wordmark. 
- Changed the vertical alignment of the service name and first line of navigation to align with the updated logo asset's wordmark baseline.
- Increased letter-spacing on product names.

## Thoughts
Alignment is very heavily affected by things like browser, operating system, the PPI of the display being used, UI scaling settings, and browser zoom settings. Much of this work, particularly alignment of text baselines, is quite likely to change with different hardware and software configurations. Tests for that are below.

Trying to maintain 7 pixels of space as a static amount is grossly impractical, as it relies on the whitespace included in the character glyph to not change. It does change. The 'P' we use in test is 7 pixels away, but curved letters like 'C' and 'O' are only 5 pixels from the wordmark boundaries. Ideally we should not be aiming for 7 pixels of perfection for every situation.

## Testing
[Testing grid](https://docs.google.com/spreadsheets/d/1RXe0c0bT9Be3lWDpUoH3ZVT3Z6Fw3IrMUl8Z0SpvL7A/edit?usp=sharing) looking at the variance in layouts across hardware and software configurations.

Chrome 135, macOS 15.3, and a 72 PPI display was the configuration used in development and is the baseline that other configurations are compared against.

CharlotteDowns commented:
I’m just looking into and asking stakeholders if we can increase the spacing between the product name and logo text, similar to that of the 3 dots width (around about 12px width)

mia-allers-gds reviewed (changes requested):
This is looking really good! To cover the points as laid out in the snags doc:

- [ ] Header height fixed
- [ ] Logo updated
- [ ] Logo vertically aligned
- [ ] Product name size adjusted to match logo cap-height
- [ ] Stacked product name padding — looks good to me, happy to go with this
- [ ] Product name letter spacing adjusted

On your comment regarding the spacing between the product name and the logo word mark, I've played around a bit. I did it just by eye, because why not. 

3px between GOV.UK and Product name:
<img width="455" alt="3px" src="https://github.com/user-attachments/assets/8f1c49ce-d359-427b-9e9f-ba676b2ddbe6" />

5px between GOV.UK and Product name:
<img width="455" alt="5px" src="https://github.com/user-attachments/assets/7b6d76a4-0e53-41ad-9101-b88fb2dc9a2f" />

@querkmachine and @CharlotteDowns please give these designs a look and let me know your thoughts.

mia-allers-gds commented:
@querkmachine @CharlotteDowns I did also notice this small thing:

The product name doesn't line up perfectly with the edge of the crown (because of the whitespace added to the first character). 

**Personally I think this is fine**, as the product name logo lockup specifies that the name should align with the word mark and we are disregarding that piece of guidance for usability reasons. But I wanted to flag it here in case anyone brings it up later.

And, actually, I quite like the alignment because it feels more aligned with the 'bulk' of the crown icon rather than its very tiny outer edge.

<img width="247" alt="Screenshot 2025-04-14 at 12 09 41" src="https://github.com/user-attachments/assets/86fff622-88ab-4df9-8c91-8e2294d97971" />

mia-allers-gds commented:
Update on the product name alignment. I'm now doubting myself and on review I think it looks kind of janky, especially when you get the underline which shoots past the left side of the text. I don't know if this is something we should try to mitigate? Could be a headache and is likely an issue with the current implementation. 

One for the backlog? What do you think @CharlotteDowns and @querkmachine ?

querkmachine commented:
@mia-allers-gds This is because of the whitespace embedded within Transport's 'P' glyph. Same reason why making there always be a specific pixel gap between the logo and product name when they're inline is so difficult—once you change what the text says, the whitespace changes with it.

This isn't a new issue, the pre-redesign header has this too, but it's less obvious because the product name is in a smaller font size and the embedded whitespace is subsequently smaller.

mia-allers-gds commented:
> @mia-allers-gds This is because of the whitespace embedded within Transport's 'P' glyph. Same reason why making there always be a specific pixel gap between the logo and product name when they're inline is so difficult—once you change what the text says, the whitespace changes with it.
> 
> This isn't a new issue, the pre-redesign header has this too, but it's less obvious because the product name is in a smaller font size and the embedded whitespace is subsequently smaller.

Yep, pretty much what I thought. We can park it/ hope nobody else notices

romaricpascal commented on packages/govuk-frontend/src/govuk/components/header/_index.scss:
```diff
@@ -134,11 +140,16 @@
 
+      // Remove 1px from the bottom to account for the font-size being 1px
+      // larger than the logo height.
+      margin-bottom: -1px;
```
**question** Do we need the value to be wrapped in `govuk-px-to-rem` as the font size is in `rem` as well?

querkmachine replied:
I'm not sure if it's really necessary—zooming scales it proportionally anyway, and alignment doesn't mean much for text scaling when the GOV.UK logo doesn't scale with it—but I've gone ahead and done it anyhow.

romaricpascal reviewed (commented):
@querkmachine Had a couple of thoughts looking at the code, so dropped a comment or two (nothing major though).

Seeing all the nudging we're having to do, I was wondering if we'd tried to align the logo to the baseline of the HTML elements rather than the HTML elements to the baseline of the logo (though this may mean that we also need to align the baseline of the different HTML elements between each other, so the same amount of nudging) 🤔

CharlotteDowns commented:
Alignment of the logo with menu button on mobile looks a little strange.
![menu button looks high on mobile view](https://github.com/user-attachments/assets/496c4145-d951-4851-8b7b-ad8b926f9cb1)

querkmachine commented on packages/govuk-frontend/src/govuk/components/header/_index.scss:
```diff
@@ -182,29 +193,12 @@
     display: inline-block;
     margin-right: govuk-spacing(2);
-    font-size: 30px; // We don't have a mixin that produces 30px font size
+    font-size: govuk-px-to-rem(31px); // We don't have a mixin that produces 30px font size
```
I noticed that this one is outside of the `_govuk-rebrand` scope so I probably shouldn't have touched it, in case it changes something unexpectedly. I'll revert it back to how it was before.

querkmachine commented:
Have made further adjustments based on our most recent go around the loop.

Note that we're back to using flexbox and `gap` for spacing the logo and product name, including the issue with hover and focus styles that introduces.

romaricpascal commented:
> Note that we're back to using flexbox and `gap` for spacing the logo and product name, including the issue with hover and focus styles that introduces.

Needs a bit of deeper cross browser testing (as well as porting to Sass rather than CSS variables), hacking in the dev tools, got something which I think is pretty close to keeping the inline hover and focus styles and giving us the right spacing with the following:

```css
.govuk-template--rebranded {

	/** We'd likely want to turn these into Sass variables rather than be magic numbers in the middle of the code */
    --header-logo-padding: 15px;
    --header-vertical-gap: 2px;
    --header-horizontal-gap: 7px;

    /** First thing, we want specific spacing around the logotype, so setting some margins around it */
    .govuk-header__logotype {
        margin-bottom: var(--header-vertical-gap);
        /** !important because I didn't take the time to figure the specificity of what was preventing this to be overriden, will need tidying up */
        margin-right: var(--header-horizontal-gap) !important;
    }

    /** To maintain the padding at the bottom of the logo, 
        we'll need to remove the vertical spacing
        as we now have a margin */
    .govuk-header__logo {
        padding-bottom: calc(var(--header-logo-padding) - var(--header-vertical-gap))
    }
    
    /** Now the fun part, let's make things shrink-wrap */
    .govuk-header__link--homepage {
        /* Retain the `display: inline` */
        display: inline;
        /* Absorb the whitespace between the logo and Product name, might need to be in `rem` rather than `px`
           or maybe `ex` so it's tied to the actual font metrics */
        word-spacing: -6px; 
        /* Then reset it in the children */
        > * {
            word-spacing: 0;
        }
        
        /* Finally tidy up the link styles, as the underline under GOV.UK overlaps the Product name a little */
		&:not(:focus) {
			/* This may need adjustments for forced colour/High contrast */
            background-color: #1f70b8;
        }
    }
}
```

querkmachine commented:
@romaricpascal I've implemented your suggestions with a couple of small tweaks. 

Not using flexbox means that whitespace characters between the logo and product name becomes significant, so I've updated the whitespace handling in Nunjucks to strip it out, and also tweaked the pixel values so that it matches the appearance it had under flexbox. 

This does mean the spacing will bee too wide if whitespace characters are still there in services, but I'm not sure if we can do all that much about that.

romaricpascal commented:
> @romaricpascal I've implemented your suggestions with a couple of small tweaks.
> 
> Not using flexbox means that whitespace characters between the logo and product name becomes significant, so I've updated the whitespace handling in Nunjucks to strip it out, and also tweaked the pixel values so that it matches the appearance it had under flexbox.
> 
> This does mean the spacing will be too wide if whitespace characters are still there in services, but I'm not sure if we can do all that much about that.

Cheers, I think even with absorbing the whitespace inside the span, there's extra spacing because of the line break between the `</svg>` and the `<span>`. This was what the `word-spacing` was set to absorb at the price of a magic number. I'm also not sure how much we can rely on the space not being there, as it's very easy for services (well, products) to miss that bit.

<img width="1512" alt="Screenshot 2025-04-22 at 17 39 13" src="https://github.com/user-attachments/assets/b2a967e6-0957-4163-8fa8-6a5d566a9922" />

The background shenanigans were to manage this overlapping of the underline of the SVG with the top of the product name.

<img width="838" alt="Screenshot 2025-04-22 at 17 39 48" src="https://github.com/user-attachments/assets/a9f57346-c8ee-419f-a037-98aa0b575db8" />

So ultimately we have a choice to make between these two options (screenshots of the `header/full-width-with-navigation` example at 1120px or so, as it's where the difference was the most visible):

<table>
<tr>
<td></td>
<th width="45%">Inline layout with <code>word-spacing</code></th>
<th width="45%">Flex layout</th>
</tr>
<tr>
<th>Default</th>
<td>
<img width="1164" alt="Screenshot 2025-04-22 at 17 52 41" src="https://github.com/user-attachments/assets/b180afd6-2cd0-410b-a091-34cb7ba51500" />
</td>
<td>
<img width="1163" alt="Screenshot 2025-04-22 at 17 52 55" src="https://github.com/user-attachments/assets/d1069ea0-ad7d-4cc6-8b2e-0b586f7b28e7" />
</td>
</tr>
<tr>
<th>Hover</th>
<td>
<img width="1165" alt="Screenshot 2025-04-22 at 17 50 18" src="https://github.com/user-attachments/assets/57309f3b-e90a-460e-8908-2dd6b2c5a935" />
(or if we're happy with the overlap)
<img width="1172" alt="Screenshot 2025-04-22 at 17 49 51" src="https://github.com/user-attachments/assets/06876f22-9377-465c-a95f-3bc534f5d674" />
</td>
<td>
<img width="1196" alt="Screenshot 2025-04-22 at 17 48 54" src="https://github.com/user-attachments/assets/dd28fbdf-1b43-4fc5-9714-69b02f3b8f81" />
</td>
</tr>
<tr>
<th>Focus</th>
<td>
<img width="1163" alt="Screenshot 2025-04-22 at 17 59 54" src="https://github.com/user-attachments/assets/cc795525-cdd8-42c0-ab29-f40d163261f5" />
</td>
<td>
<img width="1174" alt="Screenshot 2025-04-22 at 17 49 44" src="https://github.com/user-attachments/assets/da360cfe-d47a-43cd-9d5c-c989a757353a" />
</td>
</tr>
<tr>
<td>Summary</td>
<td>Focus and hover styles only wrap the logo and text tightly, at the cost of a magic number for the `word-spacing`[^1]. Extra code is needed to tidy up the hover state.</td>
<td>Focus and hover styles take the whole 33% of the space allocated for the `.govuk-header__logo` linking, taking more visual space and hover triggering when on the right of the logo. Using `margin`s like in the current state of this PR would avoid us testing for the support of `gap`.</td> 
</tr>
</table>

Feels like a good discussion for the wildcard meeting tomorrow 😊 

[^1]: Maybe there's other ways to cancel that whitespace, but I'm running out of ideas. Tried to inject a backspace character in CSS (`content: "\0008"`) with the `::before` pseudo element of `.govuk-header__product-name` but this didn't make any difference unfortunately.

querkmachine commented:
@romaricpascal Oh haha, GitHub was showing your example code in a scrollable container so I just totally missed the last half of it. 🤦

CharlotteDowns commented:
I'm leaning on **not** adjusting the padding around the logo from 15px top 13px bottom to another value, here are my reasons why:
- it moves the hover and focus state underlines closer to the bottom of the header bar
- it changes slightly depending on which browser and device our users are using
- on high contrast mode where the transparent border is visible it will look 1px closer to the bottom of the header bar

When I changed the padding to 16px top and 12px bottom there wasn't much visual difference
<img width="1062" alt="1px change to the padding to move the logo down" src="https://github.com/user-attachments/assets/ffaed646-d7da-44bb-bda8-69957b981a9e" />

When I changed the padding to 17px top and 11px bottom there was a visual difference but it did seem to become unbalanced again on the bottom side.
<img width="1063" alt="2px change to the padding to move the logo down" src="https://github.com/user-attachments/assets/37a959e9-e8d0-4564-9ed1-6a44b08487a7" />

querkmachine commented:
> I noticed a couple of things listing them out here for tomorrow's discussion.
> 
> ## Spacing between wordmark and the product name
> It looks like we still have that 1-2px gap between the wordmark and the product name. It makes the gap feel larger than intended.
> 
> ![Screenshot 2025-04-24 at 17 19 26](https://private-user-images.githubusercontent.com/112492091/437116598-b909ea78-8ed0-447c-8acb-085a4149e110.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1MTc3OTIsIm5iZiI6MTc0NTUxNzQ5MiwicGF0aCI6Ii8xMTI0OTIwOTEvNDM3MTE2NTk4LWI5MDllYTc4LThlZDAtNDQ3Yy04YWNiLTA4NWE0MTQ5ZTExMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyNFQxNzU4MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00NDgxNDU5OWY5N2ViZGZlODY0YmU0NDE5ZTg4ZDgzMmNlNzViZWYzNGY2YTk4YjdkNTBiZTczYWViODIzN2IxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.xISMS7Qp3CoGnx5-sN5HRVL0ddpMpMY0TQsLaGLNss0)
> 
> Alt: Screenshot showing gov.uk wordmark and the product name using one dot (7px) as a reference for the spacing.
> 
> ![Screenshot 2025-04-24 at 16 38 28](https://private-user-images.githubusercontent.com/112492091/437100420-1461d0a8-bd8f-4d4f-86db-cf4c115510d8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1MTc3OTIsIm5iZiI6MTc0NTUxNzQ5MiwicGF0aCI6Ii8xMTI0OTIwOTEvNDM3MTAwNDIwLTE0NjFkMGE4LWJkOGYtNGQ0Zi04NmRiLWNmNGMxMTU1MTBkOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyNFQxNzU4MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03MWYxY2ZjZDQxNzA3MzAzZjE3M2RjNjI0OTY3OTI3MGJjMzkyN2Q4ZjE3MDc2MzlmMWYwZjk2OGI1N2Q3MmJjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.3famHgdUFKlkYACAv4py6imqpRD9V24pbsstiRtle9A)
> 
> Alt: Screenshot showing gov.uk wordmark and the product name

To my knowledge we are no longer looking to have a 7px gap here, as it looked too narrow after the space between the crown icon and wordmark was increased to 12px (3 'crown dots').

As far as I remember we increased it to also be 12 pixels, and stopped taking into account the whitespace included in the Transport font glyphs as this varies depending on the initial letter: it being 'P' in our examples adds another 2 pixels to the gap.

CharlotteDowns commented on packages/govuk-frontend/src/govuk/components/header/_index.scss:
```diff
@@ -81,6 +91,11 @@
     }
+
+    @include _govuk-rebrand {
+      margin-right: 12px; // 3 'crown dots'
```
```suggestion
      margin-right: 7px; // 1 'dot'
```

I agreed with the brand team that we would have 7px space between the wordmark and the 'product name' with any additional whitespace that would come from the first letter. More info is in the [cross team decision making tab of this Google document](https://docs.google.com/document/d/1ehId5JD5Fc-pthBDlHX1mn-IwDM5Azi0375DUOsBiBk/edit?tab=t.3ha72dpcnvlv)

CharlotteDowns commented on packages/govuk-frontend/src/govuk/components/header/_index.scss:
```diff
@@ -273,7 +296,7 @@
       // Increase top/bottom padding
       padding-top: govuk-spacing(3);
-      padding-bottom: govuk-spacing(3);
+      padding-bottom: govuk-spacing(3) - $govuk-header-rebrand-logo-bottom-margin;
```
```suggestion
      padding-bottom: govuk-spacing(2) + $govuk-header-rebrand-logo-bottom-margin;
```

To get to calculated `padding-bottom: 12px;`

CharlotteDowns commented on packages/govuk-frontend/src/govuk/components/header/_index.scss:
```diff
@@ -273,7 +296,7 @@
 
       // Increase top/bottom padding
       padding-top: govuk-spacing(3);
```
```suggestion
      padding-top: govuk-spacing(3) - ($govuk-header-rebrand-logo-bottom-margin/2);
```

romaricpascal reviewed (commented):
Looks pretty much there code-wise, only two little remarks to make things more documented/easier to follow in the code. Thanks for all the attention to detail and many rounds of feedback 😊

querkmachine commented on packages/govuk-frontend/src/govuk/components/header/_index.scss:
```diff
@@ -13,11 +13,20 @@
   $govuk-header-menu-button-width: 80px;
 
+  $govuk-header-rebrand-background: $govuk-brand-colour;
+  $govuk-header-rebrand-vertical-spacing-value: 3;
```
This was to emulate the non-rebrand `$govuk-header-vertical-spacing-value` variable a few lines above it. 

I'm probably just gonna move the definitions inline rather than keeping a variable for it, as I agree that it only being used in one place kinda makes it redundant as a component-wide variable.

querkmachine commented on packages/govuk-frontend/src/govuk/components/header/_index.scss:
```diff
@@ -271,9 +303,9 @@
-      // Increase top/bottom padding
-      padding-top: govuk-spacing(3);
-      padding-bottom: govuk-spacing(3);
+      // Magic numbers, set top/bottom padding
```
Altered the comment to note that it's to attain vertical centre alignment

romaricpascal reviewed (approved):
Cheers for these final updates, looks good to merge 🙌🏻
